README
NAME
    MarpaX::Languages::ECMA::AST - Translate a ECMA source to an AST

VERSION
    version 0.001

SYNOPSIS
        use strict;
        use warnings FATAL => 'all';
        use MarpaX::Languages::ECMA::AST;
        use Log::Log4perl qw/:easy/;
        use Log::Any::Adapter;
        use Log::Any qw/$log/;
        use Data::Dumper;
        #
        # Init log
        #
        our $defaultLog4perlConf = '
        log4perl.rootLogger              = WARN, Screen
        log4perl.appender.Screen         = Log::Log4perl::Appender::Screen
        log4perl.appender.Screen.stderr  = 0
        log4perl.appender.Screen.layout  = PatternLayout
        log4perl.appender.Screen.layout.ConversionPattern = %d %-5p %6P %m{chomp}%n
        ';
        Log::Log4perl::init(\$defaultLog4perlConf);
        Log::Any::Adapter->set('Log4perl');
        #
        # Parse ECMA
        #
        my $ecmaSourceCode = 'select * from myTable;';
        my $ecmaAstObject = MarpaX::Languages::ECMA::AST->new();
        $log->infof('%s', $ecmaAstObject->parse(\$ecmaSourceCode));

DESCRIPTION
    This module translates ECMA source into an AST tree. To assist further
    process of the AST tree, the nodes of the AST are blessed according to
    the ECMA grammar you have selected. (The default is 'ECMA-262-5'.) If
    you want to enable logging, be aware that this module is using Log::Any.

SUBROUTINES/METHODS
  new($class, %options)
    Instantiate a new object. Takes as parameter an optional hash of options
    that can be:

    grammarName
        Name of a grammar. Default is 'ECMA-262-5'.

  parse($self, $sourcep)
    Get and AST from the ECMA source, pointed by $sourcep. This method will
    call all the intermediary steps (lexical, transformation, evaluation)
    necessary to produce the AST.

SEE ALSO
    Log::Any, Marpa::R2

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through the issue tracker at
    <https://rt.cpan.org/Public/Dist/Display.html?Name=MarpaX-Languages-ECMA
    -AST>. You will be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/jddurand/marpax-languages-ecma-ast>

      git clone git://github.com/jddurand/marpax-languages-ecma-ast.git

AUTHOR
    Jean-Damien Durand <jeandamiendurand@free.fr>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Jean-Damien Durand.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for MarpaX-Languages-ECMA-AST

0.001 2013-10-26T16:01:25
 [Jean-Damien Durand <jeandamiendurand@free.fr>]
 - Package ignore fix
 - Missing files for startup
 - xtest compliance
 - Program: optimizations
 - StringNumericLiteral: Removed
 - Regexp: removed
 - Marpa::R2 2.073 minimum version required
 - t/asi.t: remove STDERR prints
 - Program: fix programCompleted flag
 - Program: tracef() instead of trace()
 - t: jquery, underscore, backbone
 - Successful parsing of ECMAScript!
 - Grrr... Missing lexeme_read
 - Problem report
 - Problem report
 - Program: restore action values default
 - t/asi.t : better output
 - t: test.t deleted, asi.t created
 - Program: Fixed logging for ASI, carify code
 - Program: ASI again
 - Program: use ^INVISIBLE_SEMICOLON instead of ^INVISIBLE_SEMICOLON$
 - Program/Actions: Fake contact action
 - Postinsert space instead of newline
 - ASI new implementation
 - Grammar/ECMA_262_5/Program: Remember last S
 - Grammar/ECMA_262_5/Program: asi exception
 - Grammar/Base: keep error string if any
 - Grammar/ECMA_262_5/Program: Automatic Semicolon Insertion rule 3
 - Program: fix grammar_source initialitation
 - Systematically reserve a space at the end for an eventual semicolon
   insertion
 - Automatic Semicolon Insertion in the end callback
 - Support of an end callback
 - Grammar/ECMA_262_5/Program: apparent success of ASI for 1st case
 - Program: work on ASI
 - Base: read() and resume() are supposed to advance current position,
   protect Base against calling a subclassed method
 - Grammar/ECMA_262_5/Program: use new Base code
 - Grammar/Base : refactorisation giving more liberty to callbacks
 - Impl: added events()
 - Program: full lexemisation (2)
 - Program: full lexemisation
 - Program: Comments about this grammar difficulties
 - ECMA_262_5/Program: Removed a debug statement
 - ECMA_262_5/Program: use lexemes for masked space rules
 - Base: oups, removed debugging code
 - Identifier Ambiguity trial
 - More comments about RegularExpressionLiteral
 - Slash ambiguity
 - Base: move some logs to trace level
 - Program: Big comment explaining the pb with RegularExpressionLiteral
 - Grammar/Base: use $self in _show_last_expression
 - Grammar/Base: cache $log->is_debug and use $self->_context
 - Program: Added Missing (S) in PrimaryExpression, BreakStatement and
   ReturnStatement
 - Program: Added missing (S) in MemberExpression
 - Program: Fixed SourceElement rule
 - If error, say "last SourceElement" instead of "last expression"
 - Prepare punctuator lexemisation
 - Unified grammar in Program.pm modulo two injections
 - Fix comment magic line
 - Trying to merge everything in a single grammar
 - StringLiteral: make it more readable
 - StringLiteral: make it really independant
 - Literal: Move __UnicodeLetter to its right section
 - Program: put it comment things to remember
 - Linking program and lexical grammars, comment divided explicitely in
   singlelinet and multiline
 - Removed entry
 - Removed directory
 - Move Program grammar one level higher
 - Lexical/Program: grammar preparation
 - Grammar/Base: Support of start and length
 - Lexical: RegularExpressionLiteral injection
 - StringLiteral: fix internal rules for correct injection
 - Lexical: Injection of DecimalLiteral grammar
 - Grammar/Base.pm: generic extract method
 - Lexical/StringLiteral: fix :start rule
 - Lexical: Injection of StringLiteral grammar
 - Lexical: action RemoveUndefinedComment to get rid of comments not
   translated to a LineTerminator, so they must be discarded
 - Lexical: discard WhiteSpace
 - Lexical: when MutilineComment becomes a LineTerminator
 - Lexical: Back to :default ::= action => [values] bless => ::lhs
 - Preparation of Lexical inner String grammar
 - Removed unused dumb line
 - Removed regexp in favour of a real grammar value for Lexical items that
   need an inner grammar
 - Added missing __OctalEscapeSequence and check StringLiteral at lexing
   time
 - Removed unused code
 - Added IsZeroToThree and IsFourToSeven
 - Added strict method
 - Added IsOctalDigit, IsEight, IsNine
 - Grammar refactor + strict mode check
 - Make MultiLineComment and SingleLineComment production rules
 - eval event callbacks so that error is pretty printed
 - Fix for constaint "The source character immediately following a
   NumericLiteral must not be an IdentifierStart or DecimalDigit.": I must
   use the position
 - Change some comments
 - Factorization of character classes
 - Call the super class
 - Add $sourcep to events callback
 - The source character immediately following a NumericLiteral must not be
   an IdentifierStart or DecimalDigit.
 - Factorization
 - First commit : lexical grammar



Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Translate a ECMA source to an AST",
  "AUTHOR" => "Jean-Damien Durand <jeandamiendurand\@free.fr>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MarpaX-Languages-ECMA-AST",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "MarpaX::Languages::ECMA::AST",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Exporter" => 0,
    "Log::Any" => 0,
    "Marpa::R2" => "2.073",
    "SUPER" => 0,
    "constant" => 0,
    "diagnostics" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



